File: src/parser.rs
use serde::{Deserialize, Serialize};
use serde_yaml;

use crate::prelude::DriveType;

#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct WallConfig {
    pub endpoints: Vec<(f32, f32)>,
}

#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct StaticObjConfig {
    pub center: (f32, f32),
    pub width: f32,
    pub height: f32,
    pub rotation: f32,
}

#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct RobotConfig {
    pub id: String,
    pub pose: (f32, f32, f32),
    pub vel: (f32, f32, f32),
    pub lidar: bool,
    pub footprint: Vec<f32>,
    pub drive_type: DriveType,
    pub add_noise: bool,
}

#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct Config {
    pub robots: Vec<RobotConfig>,
    pub walls: Vec<WallConfig>,
    pub static_objects: Vec<StaticObjConfig>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct Twist {
    pub id: String,
    pub vel: (f32, f32, f32),
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct TwistArray {
    pub twists: Vec<Twist>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct Position {
    pub id: String,
    pub pose: (f32, f32, f32),
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct PositionArray {
    pub positions: Vec<Position>,
}

pub fn get_config_from_file(path: String) -> Option<Config> {
    let file_result = std::fs::File::open(path);
    match file_result {
        Ok(file) => {
            let config: Config =
                serde_yaml::from_reader(file).expect("Couldn't read config file. Rewrite properly");

            return Some(config);
        }
        Err(e) => {
            println!("Error in opening file: {}", e);
            return None;
        }
    }
}

File: bindings/xiron_py/xiron_py/parser.py
import json
from dataclasses import dataclass
from typing import Tuple, List

@dataclass
class RobotConfig:
    id: str
    pose: Tuple[float, float, float]
    vel: Tuple[float, float]
    lidar: bool = False

    def to_json(self):
        return json.dumps(self.__dict__)
    
    def from_json(self, json):
        self.id = json["id"]
        self.pose = json["pose"]
        self.vel = json["vel"]
        self.lidar = json["lidar"]

@dataclass
class Config:
    robots: List[RobotConfig]
    walls: List
    static_objects: List

    def to_json(self):
        return json.dumps(self.__dict__)
    
@dataclass
class Twist:
    id: str
    vel: Tuple[float, float]

    def to_json(self):
        return json.dumps(self.__dict__)
    
@dataclass
class TwistArray:
    twists: List[Twist]

    def to_json(self):
        my_dict = {}
        my_dict["twists"] = []
        for twist in self.twists:
            my_dict["twists"].append(json.dumps(twist.__dict__))
        
        return json.dumps(my_dict)